File: solutions/python/algorithm/others/Exercise1.py
from structure.dynamic.DynamicLinkedListADT import DynamicLinkedListADT
from structure.dynamic.DynamicQueueADT import DynamicQueueADT
from structure.dynamic.DynamicStackADT import DynamicStackADT

def combine_sorted(stack, queue):
    temp_list = DynamicLinkedListADT()
    temp_stack = DynamicStackADT()
    temp_queue = DynamicQueueADT()
    result_queue = DynamicQueueADT()

    # Extraer elementos de la pila
    while not stack.is_empty():
        value = stack.getElement()
        stack.remove()
        temp_list.add(value)
        temp_stack.add(value)

    # Restaurar pila
    while not temp_stack.is_empty():
        value = temp_stack.getElement()
        temp_stack.remove()
        stack.add(value)

    # Extraer elementos de la cola
    while not queue.is_empty():
        value = queue.getElement()
        queue.remove()
        temp_list.add(value)
        temp_queue.add(value)

    # Restaurar cola
    while not temp_queue.is_empty():
        value = temp_queue.getElement()
        temp_queue.remove()
        queue.add(value)

    # Obtener elementos, ordenarlos, y cargarlos en la nueva cola
    values = [temp_list.get(i) for i in range(temp_list.size())]
    values.sort()

    for val in values:
        result_queue.add(val)

    return result_queue



stack = DynamicStackADT()
queue = DynamicQueueADT()

# Agregar a la pila: 5 (bottom), 1, 9 (top)
stack.add(5)
stack.add(1)
stack.add(9)

# Agregar a la cola: 7 (front), 2, 3 (rear)
queue.add(7)
queue.add(2)
queue.add(3)

result = combine_sorted(stack, queue)

print("Cola resultante ordenada:")
while not result.is_empty():
    print(result.getElement(), end=" ")
    result.remove()
print()

print("Pila original:")
while not stack.is_empty():
    print(stack.getElement(), end=" ")
    stack.remove()
print()

print("Cola original:")
while not queue.is_empty():
    print(queue.getElement(), end=" ")
    queue.remove()
print()

File: solutions/javascript/src/algorithm/others/Exercise1.js
import DynamicLinkedListADT from "../../structure/dynamic/DynamicLinkedListADT.js";
import DynamicStackADT from "../../structure/dynamic/DynamicStackADT.js";
import DynamicQueueADT from "../../structure/dynamic/DynamicQueueADT.js";

function insertOrdered(list, value) {
  if (list.isEmpty()) {
    list.add(value);
    return;
  }
  let i = 0;
  while (i < list.size() && list.get(i) < value) {
    i++;
  }
  list.insert(i, value);
}

function combineSorted(stack, queue) {
  const tempList = new DynamicLinkedListADT();
  const tempStack = new DynamicStackADT();
  const tempQueue = new DynamicQueueADT();
  const resultQueue = new DynamicQueueADT();

  // Extraer elementos de pila y guardarlos en tempList ordenadamente
  while (!stack.isEmpty()) {
    const val = stack.getElement();
    stack.remove();
    insertOrdered(tempList, val);
    tempStack.add(val);
  }
  // Restaurar pila
  while (!tempStack.isEmpty()) {
    const val = tempStack.getElement();
    tempStack.remove();
    stack.add(val);
  }

  // Extraer elementos de cola y guardarlos en tempList ordenadamente
  while (!queue.isEmpty()) {
    const val = queue.getElement();
    queue.remove();
    insertOrdered(tempList, val);
    tempQueue.add(val);
  }
  // Restaurar cola
  while (!tempQueue.isEmpty()) {
    const val = tempQueue.getElement();
    tempQueue.remove();
    queue.add(val);
  }

  // Agregar ordenadamente a la cola resultado
  for (let i = 0; i < tempList.size(); i++) {
    resultQueue.add(tempList.get(i));
  }

  return resultQueue;
}


const stack = new DynamicStackADT();
const queue = new DynamicQueueADT();

// Datos para pila
stack.add(5);
stack.add(1);
stack.add(9);

// Datos para cola
queue.add(7);
queue.add(2);
queue.add(3);

const resultQueue = combineSorted(stack, queue);

console.log("Cola resultante ordenada:");
while (!resultQueue.isEmpty()) {
    console.log(resultQueue.getElement());
    resultQueue.remove();
}

console.log("Pila original:");
while (!stack.isEmpty()) {
    console.log(stack.getElement());
    stack.remove();
}

console.log("Cola original:");
while (!queue.isEmpty()) {
    console.log(queue.getElement());
    queue.remove();
}